// @andyet/simplewebrtc/Constants.js
"use strict";
// Action Names
// ====================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.SET_USER_PREFERENCE = '@andyet/SET_USER_PREFERENCE';
exports.JOIN_ROOM = '@andyet/JOIN_ROOM';
exports.LEAVE_ROOM = '@andyet/LEAVE_ROOM';
exports.LOCK_ROOM = '@andyet/LOCK_ROOM';
exports.UNLOCK_ROOM = '@andyet/UNLOCK_ROOM';
exports.DESTROY_ROOM = '@andyet/DESTROY_ROOM';
exports.SELF_UPDATED = '@andyet/SELF_UPDATED';
exports.JOIN_ROOM_FAILED = '@andyet/JOIN_ROOM_FAILED';
exports.JOIN_ROOM_SUCCESS = '@andyet/JOIN_ROOM_SUCCESS';
exports.ROOM_LOCKED = '@andyet/ROOM_LOCKED';
exports.ROOM_UNLOCKED = '@andyet/ROOM_UNLOCKED';
exports.ROOM_DESTROYED = '@andyet/ROOM_DESTROYED';
exports.KICK_PEER = '@andyet/KICK_PEER';
exports.PEER_ONLINE = '@andyet/PEER_ONLINE';
exports.PEER_OFFLINE = '@andyet/PEER_OFFLINE';
exports.PEER_UPDATED = '@andyet/PEER_UPDATED';
exports.CHAT_INCOMING = '@andyet/CHAT_INCOMING';
exports.CHAT_OUTGOING = '@andyet/CHAT_OUTGOING';
exports.CHAT_STATE_OUTGOING = '@andyet/CHAT_STATE_OUTGOING';
exports.RTT_INCOMING = '@andyet/RTT_INCOMING';
exports.RTT_OUTGOING = '@andyet/RTT_OUTGOING';
exports.ADD_MEDIA = '@andyet/ADD_MEDIA';
exports.REMOVE_MEDIA = '@andyet/REMOVE_MEDIA';
exports.MEDIA_UPDATED = '@andyet/MEDIA_UPDATED';
exports.SIGNALING_CLIENT = '@andyet/SIGNALING_CLIENT';
exports.SIGNALING_CLIENT_SHUTDOWN = '@andyet/SIGNALING_CLIENT_SHUTDOWN';
exports.CONNECTION_STATE_CHANGE = '@andyet/CONNECTION_STATE_CHANGE';
exports.RECEIVED_CONFIG = '@andyet/RECEIVED_CONFIG';
exports.RECEIVED_CONFIG_ERROR = '@andyet/RECEIVED_CONFIG_ERROR';
exports.QUEUE_TELEMETRY = '@andyet/QUEUE_TELEMETRY';
exports.TELEMETRY_SUCCESS = '@andyet/TELEMETRY_SUCCESS';
exports.JOIN_CALL = '@andyet/JOIN_CALL';
exports.LEAVE_CALL = '@andyet/LEAVE_CALL';
exports.SET_CALL_PREFERENCE = '@andyt/SET_CALL_PREFERENCE';
exports.PEER_CONNECTION_UPDATED = '@andyet/PEER_CONNECTION_UPDATED';
exports.PEER_CONNECTION_ADDED = '@andyet/PEER_CONNECTION_ADDED';
exports.PEER_CONNECTION_REMOVED = '@andyet/PEER_CONNECTION_REMOVED';
exports.DEVICES = '@andyet/DEVICES';
exports.CAMERA_PERMISSION_DENIED = '@andyet/CAMERA_PERMISSION_DENIED';
exports.MICROPHONE_PERMISSION_DENIED = '@andyet/MICROPHONE_PERMISSION_DENIED';
exports.DEVICE_CAPTURE = '@andyet/DEVICE_CAPTURE';
exports.SET_VIDEO_RESOLUTION_TIERS = '@andyet/SET_VIDEO_RESOLUTION_TIERS';
// Constants
// ====================================================================
exports.SDK_VERSION = '1.22.1';
exports.DIRECTION_INCOMING = 'incoming';
exports.DIRECTION_OUTGOING = 'outgoing';
